motion_computation/src/psm_to_external_object_convertor.cpp

```diff
@@ -0,0 +1,501 @@
+
+  RCLCPP_DEBUG_STREAM(rclcpp::get_logger("motion_computation::conversion"), "R_n_h (x,y,z,w) : ( " << R_n_h.x() << ", " << R_n_h.y() << ", " << R_n_h.z() << ", " << R_n_h.w());
+
+  RCLCPP_DEBUG_STREAM(rclcpp::get_logger("motion_computation::conversion"), "R_m_s (x,y,z,w) : ( " << R_m_s.x() << ", " << R_m_s.y() << ", " << R_m_s.z() << ", " << R_m_s.w());
+
+  tf2::Transform T_m_s(R_m_s, tf2::Vector3(map_point.x(), map_point.y(),
+                                           map_point.z()));  // Reported position and orientation
+                                                             // of sensor frame in map frame
+
+        tf2::Transform T_m_n_no_heading(R_m_n, tf2::Vector3(0, 0, 0)); // Used to transform the covariance
+
+        // This covariance represents the covariance of the NED frame N-lat,
+        // E-lon, heading- angle east of north This means that the covariance is
+        // in the NED frame, and needs to be transformed to the map frame
+        // clang-format off
+        std::array<double, 36> input_covariance = { 
+          lat_variance, 0, 0,  0, 0, 0,
+          0, lon_variance, 0,  0, 0, 0,
+          0, 0, 0,  0, 0, 0,
+          0,  0,  0,  1,  0, 0,
+          0,  0,  0,  0,  1, 0, 
+          0,  0,  0,  0,  0, heading_variance
+        };
+        // clang-format on
+
+        std::array<double, 36> new_cov =
+            tf2::transformCovariance(input_covariance,
+                                     // Per the usage of transformCovariance in tf2_geometry_msgs
+                                     // this frame should be the transform between the map which the pose
+                                     // is in an the target frame.
+                                     T_m_n_no_heading);
+
+        // Populate message
+        geometry_msgs::msg::PoseWithCovariance pose;
+        pose.pose.position.x = T_m_s.getOrigin().getX();
+        pose.pose.position.y = T_m_s.getOrigin().getY();
+        pose.pose.position.z = T_m_s.getOrigin().getZ();
+
+        pose.pose.orientation.x = T_m_s.getRotation().getX();
+        pose.pose.orientation.y = T_m_s.getRotation().getY();
+        pose.pose.orientation.z = T_m_s.getRotation().getZ();
+        pose.pose.orientation.w = T_m_s.getRotation().getW();
+
+        pose.covariance = new_cov;
+
+        return pose;
+}
+
+std::vector<carma_perception_msgs::msg::PredictedState> predicted_poses_to_predicted_state(
+    const std::vector<geometry_msgs::msg::Pose>& poses, double constant_velocity, const rclcpp::Time& start_time,
+    const rclcpp::Duration& step_size, const std::string& frame, double initial_pose_confidence,
+    double initial_vel_confidence) {
+  std::vector<carma_perception_msgs::msg::PredictedState> output;
+  output.reserve(poses.size());
+
+  rclcpp::Time time(start_time);
+
+  for (auto p : poses) {
+    time += step_size;
```

Comment:
We're incrementing the time as soon as the loop starts, which means there is no predicted state for 0 time. Which makes sense, but wanted to check if that was the intention

Reply:
Yeah thats intentional. These are just future predictions